### PR TITLE
Issue 257: Enable reader offline API for ReaderGroup

### LINKED ISSUE #257: Unable to manually mark a stopped reader as offline
**Problem description**
When you don't explicitly close a reader it will be stuck forever in an online state because it's never detected as failed and there are no methods to close it without creating a reader.

Sample code:
``` python
reader_group = manager.create_reader_group(readergroup, scope, stream)
reader = reader_group.create_reader(reader)
```
thread '<unnamed>' panicked at 'Error while creating the reader: SyncError { error_msg: "add reader Reader { name: \"<reader>\" }", source: SyncUpdateError { error_msg: "Failed to add online reader Reader { name: \"reader\" }: reader already online" } }

**Problem location**
python client binding

**Suggestions for an improvement**
Implement methods to list all readers in a reader group and be able to close them by name/id.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
ReaderGroup can mark a reader as offline. 

**Purpose of the change**  
Fixes #257 

**What the code does**  
* This enables users to invoke a reader offline manually with the help of a reader group.
* The public APIs for acquire_segment, reader_offline, release_segment, release_segment_at have been changed.
* The error handling behavior of the APIs has been improved.

**How to verify it**  
All the existing and newly added tests should pass.

### CODE REVIEW

File: src/event/reader.rs [hunk 1]
```diff
@@ -1545,7 +1548,11 @@ mod tests {
         );
 
         // acquire a segment
-        let mut slice = cf.runtime().block_on(reader.acquire_segment()).unwrap();
+        let mut slice = cf
+            .runtime()
+            .block_on(reader.acquire_segment())
+            .expect("Failed to acquire a segment")
```
Comment (tkaitchuck) [inline src/event/reader.rs:1554]: same here

File: src/event/reader.rs [hunk 2]
```diff
@@ -1554,16 +1561,24 @@ mod tests {
...
         // release the segment slice.
-        cf.runtime().block_on(reader.release_segment(slice));
+        let _ = cf.runtime().block_on(reader.release_segment(slice));
 
         // acquire the next segment
-        let slice = cf.runtime().block_on(reader.acquire_segment()).unwrap();
+        let slice = cf
+            .runtime()
+            .block_on(reader.acquire_segment())
+            .expect("Failed to acquire segment")
```
Comment (tkaitchuck) [inline src/event/reader.rs:1570]: same here and below

File: examples/event_write_and_read.rs [hunk 3]
```diff
@@ -78,12 +78,19 @@ fn main() {
         println!("event reader created");
 
         // read from segment
-        let mut slice = reader.acquire_segment().await.expect("acquire segment");
+        let mut slice = reader
+            .acquire_segment()
+            .await
+            .unwrap()
```
[the remark this answers is not in the document]
Comment (shrids) [reply examples/event_write_and_read.rs:84]: I have refactored the code here.

File: src/event/reader.rs [hunk 4]
```diff
@@ -86,13 +86,13 @@ cfg_if::cfg_if! {
 ///     // readers which are part of the reader group.
 ///     let mut reader1 = rg.create_reader("r1".to_string()).await;
 ///     // read all events from a given segment slice.
-///     if let Some(mut segment_slice) =  reader1.acquire_segment().await {
+///     if let Some(mut segment_slice) =  reader1.acquire_segment().await.expect("Failed to acquire segment") {
```
[the remark this answers is not in the document]
Comment (shrids) [reply src/event/reader.rs:89]: improved the msg.

File: src/event/reader.rs (hunk 1 above)
Comment (shrids) [reply src/event/reader.rs:1554]: fixed it.

File: src/event/reader.rs (hunk 2 above)
Comment (shrids) [reply src/event/reader.rs:1570]: fixed it.

File: src/event/reader.rs [hunk 5]
```diff
@@ -686,12 +764,16 @@ impl ReaderState {
     async fn wait_for_segment_slice_return(&mut self, segment: &ScopedSegment) -> Option<SliceMetadata> {
         if let Some(receiver) = self.slice_release_receiver.remove(segment) {
             match receiver.await {
+                //TODO: Handle error here.
```
Comment (shrids) [inline src/event/reader.rs:767]: Returning a none incase of an Error should be fine here. (Testing of this part is in progress).

File: src/event/reader.rs [hunk 6]
```diff
@@ -400,21 +441,38 @@ impl EventReader {
     /// to different Segments of the stream are received. In-case we receive data for an already
     /// acquired SegmentSlice this method waits until SegmentSlice is completely consumed before
     /// returning the data.
-    pub async fn acquire_segment(&mut self) -> Option<SegmentSlice> {
-        info!("acquiring segment for reader {}", self.id);
+    ///
+    /// Note: it may return an error indicating that the reader is not online. This means
+    /// that another thread removes this reader from the ReaderGroup probably because the host of this reader
+    /// is assumed dead.
+    pub async fn acquire_segment(&mut self) -> Result<Option<SegmentSlice>, Error> {
```
Comment (tkaitchuck) [inline src/event/reader.rs:448]: Humm, should we make our own return value type? Also is io.Error really what we want here? It makes sense for byteStream, but here I am less sure.

File: src/event/reader.rs [hunk 7]
```diff
@@ -1038,6 +1104,18 @@ impl SliceMetadata {
     pub fn has_events(&self) -> bool {
         !self.partial_data_present && self.segment_data.value.len() > TYPE_PLUS_LENGTH_SIZE as usize
     }
+
+    fn copy_meta(&self) -> SliceMetadata {
```
Comment (tkaitchuck) [inline src/event/reader.rs:1108]: Maybe implement the clone trait?

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

File: integration_test/src/reader_group_tests.rs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        }
+    };
+    // Segment slice is dropped here and it will update the RG state with the offsets.
+    // Now mark the reader offline
+    let offline_res = h.block_on(rg.reader_offline(reader_id.to_string(), None));
+    assert!(offline_res.is_ok(), "Reader offline did not succeed");
+
+    // Attempt reading the remaining events.
+    while let Some(slice) = h
+        .block_on(reader1.acquire_segment())
```
[the remark this answers is not in the document]
Comment (shrids) [reply integration_test/src/reader_group_tests.rs]: fixed this.

File: src/event/reader.rs (hunk 6 above)
Comment (shrids) [reply src/event/reader.rs:460]: I have improved the code and introduced a custom error called `EventReaderError`

File: src/event/reader.rs (hunk 7 above)
Comment (shrids) [reply src/event/reader.rs:1126]: This method only copies the Segment metadata and not the segment Data Buffer. Hence I do not think we should implement the clone trait.

File: src/event/reader.rs [hunk 9]
```diff
@@ -389,8 +442,17 @@ impl EventReader {
                 .await
                 .release_segment(&self.id, &segment, &Offset::new(read_offset))
                 .await
-                .expect("Failed to release segment from RG state for reader");
+                .map_err(|e| {
+                    Error::new(
```
Comment (tkaitchuck) [inline src/event/reader.rs:446]: The meaning of this is not clear. It looks here as though this is a catch all error. But the docs of the caller are explicet that the reason for an error is that the reader was already offline when offline or release was invoked.

Comment (shrids) [reply src/event/reader.rs:446]: fixed this and reader offline API as well.

**Review by tkaitchuck — APPROVED** [review]

File: python_binding/src/stream_reader.rs [hunk 10]
```diff
@@ -86,7 +99,13 @@ impl StreamReader {
     pub fn release_segment(&self, slice: &mut Slice) -> PyResult<()> {
         info!("Release segment slice back");
         if let Some(s) = slice.get_set_to_none() {
-            self.runtime_handle.block_on(self.release_segment_async(s));
+            match self.runtime_handle.block_on(self.release_segment_async(s)) {
```
[the remark this answers is not in the document]
Comment (shrids) [reply python_binding/src/stream_reader.rs]: thanks, I have improved the code.

**Pull request merged**